File: lib/features/analytics/domain/entities/storage_breakdown.dart
import 'package:equatable/equatable.dart';

class StorageBreakdown extends Equatable {
  final String packageName;

  final int apkSize;

  final int codeSize;

  final int appDataInternal;

  final int cacheInternal;

  final int cacheExternal;

  final int obbSize;

  final int externalDataSize;

  final int mediaSize;

  final int databasesSize;

  final int logsSize;

  final int residualSize;

  final MediaBreakdown mediaBreakdown;

  final Map<String, int> databaseBreakdown;

  final int totalExact;

  final int totalEstimated;

  final int totalCombined;

  final DateTime scanTimestamp;

  final double confidenceLevel;

  final List<String> limitations;

  const StorageBreakdown({
    required this.packageName,
    this.apkSize = 0,
    this.codeSize = 0,
    this.appDataInternal = 0,
    this.cacheInternal = 0,
    this.cacheExternal = 0,
    this.obbSize = 0,
    this.externalDataSize = 0,
    this.mediaSize = 0,
    this.databasesSize = 0,
    this.logsSize = 0,
    this.residualSize = 0,
    this.mediaBreakdown = const MediaBreakdown(),
    this.databaseBreakdown = const {},
    this.totalExact = 0,
    this.totalEstimated = 0,
    this.totalCombined = 0,
    required this.scanTimestamp,
    this.confidenceLevel = 0.0,
    this.limitations = const [],
  });

  factory StorageBreakdown.fromMap(Map<Object?, Object?> map) {
    final mediaBreakdownMap = map['mediaBreakdown'] as Map<Object?, Object?>?;
    final dbBreakdownMap = map['databaseBreakdown'] as Map<Object?, Object?>?;

    return StorageBreakdown(
      packageName: map['packageName'] as String? ?? '',
      apkSize: map['apkSize'] as int? ?? 0,
      codeSize: map['codeSize'] as int? ?? 0,
      appDataInternal: map['appDataInternal'] as int? ?? 0,
      cacheInternal: map['cacheInternal'] as int? ?? 0,
      cacheExternal: map['cacheExternal'] as int? ?? 0,
      obbSize: map['obbSize'] as int? ?? 0,
      externalDataSize: map['externalDataSize'] as int? ?? 0,
      mediaSize: map['mediaSize'] as int? ?? 0,
      databasesSize: map['databasesSize'] as int? ?? 0,
      logsSize: map['logsSize'] as int? ?? 0,
      residualSize: map['residualSize'] as int? ?? 0,
      mediaBreakdown: mediaBreakdownMap != null
          ? MediaBreakdown.fromMap(mediaBreakdownMap)
          : const MediaBreakdown(),
      databaseBreakdown:
          dbBreakdownMap?.map(
            (key, value) => MapEntry(key.toString(), value as int),
          ) ??
          {},
      totalExact: map['totalExact'] as int? ?? 0,
      totalEstimated: map['totalEstimated'] as int? ?? 0,
      totalCombined: map['totalCombined'] as int? ?? 0,
      scanTimestamp: DateTime.fromMillisecondsSinceEpoch(
        map['scanTimestamp'] as int? ?? DateTime.now().millisecondsSinceEpoch,
      ),
      confidenceLevel: (map['confidenceLevel'] as num?)?.toDouble() ?? 0.0,
      limitations:
          (map['limitations'] as List<Object?>?)
              ?.map((e) => e.toString())
              .toList() ??
          [],
    );
  }

  int get totalCache => cacheInternal + cacheExternal;

  bool get isDetailed => totalEstimated > 0 || limitations.isNotEmpty;

  bool get isRecent {
    final age = DateTime.now().difference(scanTimestamp);
    return age.inMinutes < 5;
  }

  @override
  List<Object?> get props => [
    packageName,
    apkSize,
    codeSize,
    appDataInternal,
    cacheInternal,
    cacheExternal,
    obbSize,
    externalDataSize,
    mediaSize,
    databasesSize,
    logsSize,
    residualSize,
    mediaBreakdown,
    databaseBreakdown,
    totalExact,
    totalEstimated,
    totalCombined,
    scanTimestamp,
    confidenceLevel,
    limitations,
  ];
}

class MediaBreakdown extends Equatable {
  final int images;
  final int videos;
  final int audio;
  final int documents;

  const MediaBreakdown({
    this.images = 0,
    this.videos = 0,
    this.audio = 0,
    this.documents = 0,
  });

  factory MediaBreakdown.fromMap(Map<Object?, Object?> map) {
    return MediaBreakdown(
      images: map['images'] as int? ?? 0,
      videos: map['videos'] as int? ?? 0,
      audio: map['audio'] as int? ?? 0,
      documents: map['documents'] as int? ?? 0,
    );
  }

  int get total => images + videos + audio + documents;

  @override
  List<Object?> get props => [images, videos, audio, documents];
}

File: android/app/src/main/kotlin/com/rakhul/unfilter/StorageBreakdown.kt
package com.rakhul.unfilter

/**
 * Comprehensive storage breakdown for an application.
 * Provides granular insight into where storage is consumed.
 */
data class StorageBreakdown(
    // ===== EXACT MEASUREMENTS (from StorageStatsManager) =====
    
    /** APK file size + split APKs */
    val apkSize: Long = 0L,
    
    /** Code size (DEX, native libs inside APK) */
    val codeSize: Long = 0L,
    
    /** Internal app data (databases, shared_prefs, etc.) */
    val appDataInternal: Long = 0L,
    
    /** Internal cache directory */
    val cacheInternal: Long = 0L,
    
    /** External cache directory */
    val cacheExternal: Long = 0L,
    
    // ===== DISCOVERABLE (via file analysis) =====
    
    /** OBB files size (Android/obb/{package}) */
    val obbSize: Long = 0L,
    
    /** External app data (Android/data/{package}) */
    val externalDataSize: Long = 0L,
    
    /** Media files owned/created by app (images, videos, audio) */
    val mediaSize: Long = 0L,
    
    /** Database files total size */
    val databasesSize: Long = 0L,
    
    /** Log files (.log, .txt logs) */
    val logsSize: Long = 0L,
    
    /** Other discoverable files not categorized */
    val residualSize: Long = 0L,
    
    // ===== DETAILED BREAKDOWNS =====
    
    /** Media breakdown by type */
    val mediaBreakdown: MediaBreakdown = MediaBreakdown(),
    
    /** Individual database files: "main.db" -> 54MB */
    val databaseBreakdown: Map<String, Long> = emptyMap(),
    
    // ===== METADATA =====
    
    /** Sum of exact measurements from official APIs */
    val totalExact: Long = 0L,
    
    /** Sum of estimated/discovered measurements */
    val totalEstimated: Long = 0L,
    
    /** Grand total (exact + estimated) */
    val totalCombined: Long = 0L,
    
    /** When this breakdown was computed */
    val scanTimestamp: Long = System.currentTimeMillis(),
    
    /** Confidence level (0.0 - 1.0) based on API availability and access permissions */
    val confidenceLevel: Float = 0.0f,
    
    /** List of limitations encountered during analysis */
    val limitations: List<String> = emptyList(),
    
    /** Package name this breakdown belongs to */
    val packageName: String = ""
) {
    /**
     * Convert to Map for platform channel transmission.
     * Avoids TransactionTooLargeException by keeping payload reasonable.
     */
    fun toMap(): Map<String, Any?> {
        return mapOf(
            "packageName" to packageName,
            "apkSize" to apkSize,
            "codeSize" to codeSize,
            "appDataInternal" to appDataInternal,
            "cacheInternal" to cacheInternal,
            "cacheExternal" to cacheExternal,
            "obbSize" to obbSize,
            "externalDataSize" to externalDataSize,
            "mediaSize" to mediaSize,
            "databasesSize" to databasesSize,
            "logsSize" to logsSize,
            "residualSize" to residualSize,
            "mediaBreakdown" to mediaBreakdown.toMap(),
            "databaseBreakdown" to databaseBreakdown,
            "totalExact" to totalExact,
            "totalEstimated" to totalEstimated,
            "totalCombined" to totalCombined,
            "scanTimestamp" to scanTimestamp,
            "confidenceLevel" to confidenceLevel,
            "limitations" to limitations
        )
    }
    
    companion object {
        /**
         * Create a minimal breakdown for cases where full analysis fails.
         */
        fun minimal(packageName: String, basicSize: Long): StorageBreakdown {
            return StorageBreakdown(
                packageName = packageName,
                apkSize = basicSize,
                totalExact = basicSize,
                totalCombined = basicSize,
                confidenceLevel = 0.3f,
                limitations = listOf("Unable to perform detailed analysis")
            )
        }
    }
}

/**
 * Media storage breakdown by type.
 */
data class MediaBreakdown(
    val images: Long = 0L,
    val videos: Long = 0L,
    val audio: Long = 0L,
    val documents: Long = 0L
) {
    fun toMap(): Map<String, Long> {
        return mapOf(
            "images" to images,
            "videos" to videos,
            "audio" to audio,
            "documents" to documents
        )
    }
    
    val total: Long
        get() = images + videos + audio + documents
}

/**
 * Cached result with expiration.
 */
data class CachedBreakdown(
    val breakdown: StorageBreakdown,
    val cachedAt: Long = System.currentTimeMillis()
) {
    companion object {
        const val CACHE_VALIDITY_MS = 5 * 60 * 1000L // 5 minutes
    }
    
    fun isValid(): Boolean {
        return (System.currentTimeMillis() - cachedAt) < CACHE_VALIDITY_MS
    }
}
